### PR TITLE
Asset Pair Manager V2

### PR DESCRIPTION
**Pull request by richardliang**

### CODE REVIEW

File: contracts/managers/lib/FlexibleTimingManagerLibrary.sol [hunk 1]
```diff
@@ -59,6 +60,32 @@ library FlexibleTimingManagerLibrary {
         );        
     }
 
+    /*
```
Comment (bweick) [inline contracts/managers/lib/FlexibleTimingManagerLibrary.sol:63]: Is this necessary? There's no real logic change is there?

File: contracts/managers/AssetPairManagerV2.sol [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+        return trigger.isBullish() ? bullishBaseAssetAllocation : bearishBaseAssetAllocation;  
+    }
+
+     /*
+     * Function returning whether the rebalanceInterval has elapsed and then RebalancingSetToken is in 
+     * Default state
+     *
+     * @return       Whether RebalancingSetToken is in valid state for rebalance
+     */
+    function rebalancingSetTokenInValidState()
```
Comment (bweick) [inline contracts/managers/AssetPairManagerV2.sol:357]: One thing we could consider is getting rid of the `FlexibleTimingManagerLibrary` and just use this. Think the reason we didn't before was for more robust error logging but not sure that's worth it/necessary.

File: test/contracts/managers/assetPairManagerV2.spec.ts [hunk 3]
```diff
@@ -0,0 +1,1735 @@
...
+    beforeEach(async () => {
+      subjectCore = core.address;
+      subjectAllocator = allocator.address;
+      subjectTrigger = trigger.address;
+      subjectUseBullishAllocation = true;
+      subjectAllocationDenominator = new BigNumber(100);
+      subjectBullishBaseAssetAllocation = new BigNumber(100);
+      subjectSignalConfirmationMinTime = ONE_HOUR_IN_SECONDS.mul(6);
+      subjectSignalConfirmationMaxTime = ONE_HOUR_IN_SECONDS.mul(12);
+      subjectLiquidatorData = ZERO_BYTES;
```
Comment (bweick) [inline test/contracts/managers/assetPairManagerV2.spec.ts:292]: Can we make this something not `ZERO_BYTES` just to be sure it isn't set. Shouldn't affect the ability to successfully rebalance cause the current liquidator just ignores the `liquidatorData` parma

File: test/contracts/managers/assetPairManagerV2.spec.ts [hunk 4]
```diff
@@ -0,0 +1,1735 @@
...
+
+      it('should revert', async () => {
+        await expectRevertError(subject());
+      });
+    });
+
+    describe('but a rebalancing set has already been initialized', async () => {
+      it('should revert', async () => {
+        // Initialize the first time
+        await subject();
```
Comment (bweick) [inline test/contracts/managers/assetPairManagerV2.spec.ts:522]: Do this in a `beforeEach` block

File: contracts/managers/AssetPairManagerV2.sol (hunk 2 above)
Comment (richardliang) [reply contracts/managers/AssetPairManagerV2.sol:357]: This way sounds good to me

File: test/contracts/managers/assetPairManagerV2.spec.ts [hunk 5]
```diff
@@ -0,0 +1,1735 @@
...
+        subjectCaller = attackerAccount;
+      });
+
+      it('should revert', async () => {
+        await expectRevertError(subject());
+      });
+    });
+  });
+
+  describe('#adjustFee', async () => {
```
Comment (bweick) [inline test/contracts/managers/assetPairManagerV2.spec.ts:1111]: Let's write a quick cancel test case as well

File: contracts/managers/AssetPairManagerV2.sol [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+ * @author Set Protocol
+ *
+ * Manager contract for implementing any trading pair and strategy for RebalancingSetTokenV3. Allocation 
+ * determinations are made based on output of Trigger contract. bullishBaseAssetAllocation amount is 
+ * passed in and used when bullish, allocationDenominator - bullishBaseAssetAllocation used when bearish.
+ * 
+ * CHANGELOG:
+ * - Support RebalancingSetTokenV3
+ * - Remove logic associated with pricing auctions, which has been moved to liquidator contracts
+ * - Add abilities to switch liquidator and liquidatorData, and ability to change fees
```
Comment (felix2feng) [inline contracts/managers/AssetPairManagerV2.sol:43]: We may also want to change the feeRecipient too

Comment (felix2feng) [reply contracts/managers/AssetPairManagerV2.sol]: Ah i see you did it. I suppose, just add it to the changelog then

File: utils/helpers/protocolHelper.ts [hunk 7]
```diff
@@ -253,6 +290,55 @@ export class ProtocolHelper {
     );
   }
 
+  public async createDefaultRebalancingSetTokenV3Async(
```
Comment (felix2feng) [inline utils/helpers/protocolHelper.ts:293]: Does this helper not exist on `set-protocol-contracts`?

File: contracts/managers/AssetPairManagerV2.sol (hunk 6 above)
Comment (richardliang) [reply contracts/managers/AssetPairManagerV2.sol:43]: setFeeRecipient is already built in. Going to clarify this is the comment here

File: utils/helpers/protocolHelper.ts (hunk 7 above)
Comment (richardliang) [reply utils/helpers/protocolHelper.ts:293]: Yeah it does. But, there was an issue importing this function from contracts. Similar issue to the one Brian ran into while importing this into setprotocol.js a few weeks ago

**Pull request merged**